components/script/task_queue.rs

```diff
@@ -81,20 +87,16 @@ impl<T: QueuedTaskConversion> TaskQueue<T> {
 
         for msg in to_be_throttled {
             // Categorize tasks per task queue.
-            let (worker, category, boxed, pipeline_id) = match msg.into_queued_task() {
-                Some((worker, category, boxed, pipeline_id)) => (worker, category, boxed, pipeline_id),
+            let (worker, category, boxed, pipeline_id, task_source) = match msg.into_queued_task() {
+                Some(queued_task) =>
+                    queued_task,
                 None => unreachable!("A message to be throttled should always be convertible into a queued task"),
             };
-            // FIXME: Add the task-source name directly to CommonScriptMsg::Task.
-            let task_source = match category {
-                ScriptThreadEventCategory::PerformanceTimelineTask => TaskSourceName::PerformanceTimeline,
-                _ => unreachable!(),
-            };
             let mut throttled_tasks = self.throttled.borrow_mut();
             throttled_tasks
-                .entry(task_source)
+                .entry(task_source.clone())
```

Comment:
I'm not sure this is a good idea.

I thought about refactoring this to use references instead of owned objects, but I didn't felt brave enough and went the easy way.

Any suggestion on what would be better here?

Reply:
Yeah, that was my intuition here (i.e. the `entry` call was worth a clone in an small enum). Thank you for the detailed explanation!